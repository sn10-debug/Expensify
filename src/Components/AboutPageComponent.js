import React from "react";
export default class AboutPageComponent extends React.Component {
  render() {
    return (
      <div>
        <p>This is a About Page</p>
      </div>
    );
  }
}
