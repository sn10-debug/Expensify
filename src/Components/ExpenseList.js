import React from "react";
import { connect } from "react-redux";
import Expense from "./Expense";
import filterExpenses from "../selectors/expenses";

export class ExpenseList extends React.Component {
  render() {
    return (
      <div>
        <p>Expense List : </p>

        {this.props.expenses.map((mov) => (
          <Expense
            history={this.props.history}
            key={mov.id}
            {...mov}
            edit={this.props.edit}
          />
        ))}
      </div>
    );
  }
}

let ConnectedExpenseList = connect((state) => {
  // In this function we have access to the state of the store
  return {
    expenses: filterExpenses(state.expenses, state.filters),
  };
  //   Now this object will be passed as a prop to the component
})(ExpenseList);

export default ConnectedExpenseList;
