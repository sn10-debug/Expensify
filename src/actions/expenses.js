import moment from "moment";
import { v4 as uuidv4 } from "uuid";

let defaultDescription =
  "Lorem ipsum dolor sit amet consectetur adipisicing elit. Nisi et iusto magnam commodi dolores obcaecati est aut placeat deserunt illo sequi aspernatur ipsam labore quidem distinctio, laborum maiores enim iste?";

class expense {
  constructor(
    name,
    description = defaultDescription,
    amount,
    createDate = moment()
  ) {
    this.id = uuidv4();
    this.name = name;
    this.description = description;
    this.amount = amount;
    this.createDate = createDate;
  }
}

export const addExpense = ({
  name,
  description = defaultDescription,
  amount,
  date = moment(),
}) => ({
  type: "ADD_EXPENSE",
  expense: new expense(name, description, amount, date),
});

export const removeExpense = ({ id }) => {
  return {
    type: "REMOVE_EXPENSE",
    id,
  };
};

export const editExpense = ({ id, changes }) => {
  return {
    type: "EDIT_EXPENSE",
    id,
    changes,
  };
};
