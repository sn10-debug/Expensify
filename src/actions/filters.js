export const setTextfilter = (text = "") => ({ type: "SET_TEXT", text });
export const setSortBy = (method = "date") => ({
  type: "SET_SORT_METHOD",
  method,
});
export const setStartDate = ({ date = undefined } = {}) => {
  return {
    type: "SET_START_DATE",
    date,
  };
};
export const setEndDate = ({ date = undefined } = {}) => {
  return {
    type: "SET_END_DATE",
    date,
  };
};
