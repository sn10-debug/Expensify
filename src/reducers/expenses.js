export default (state = [], action) => {
  switch (action.type) {
    case "ADD_EXPENSE":
      return [...state, action.expense];
    case "REMOVE_EXPENSE":
      return state.filter((mov) => mov.id !== action.id);

    case "EDIT_EXPENSE":
      return state.map((mov) => {
        if (mov.id === action.id) return { ...mov, ...action.changes };
        else return mov;
      });
    default:
      return state;
  }
};
