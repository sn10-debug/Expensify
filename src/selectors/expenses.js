const filterExpenses = (expenses, { text, sortBy, startDate, endDate }) => {
  let filteredExpenses = expenses.filter((expense) => {
    let dateMatch, textMatch;

    if (
      (startDate
        ? expense.createDate.valueOf() >= startDate.valueOf()
        : true) &&
      (endDate ? expense.createDate.valueOf() <= endDate.valueOf() : true)
    )
      dateMatch = true;

    if (expense.description.toLowerCase().split(text.toLowerCase()).length > 1)
      textMatch = true;

    return dateMatch && textMatch;
  });

  if (sortBy === "date") {
    filteredExpenses.sort(
      (a, b) =>
        new Date(new Date() - a.createDate.valueOf()) -
        new Date(new Date() - b.createDate.valueOf())
    );
    return filteredExpenses;
  }

  if (sortBy === "amount") {
    filteredExpenses.sort((a, b) => b.amount - a.amount);
    return filteredExpenses;
  }
};

export default filterExpenses;
