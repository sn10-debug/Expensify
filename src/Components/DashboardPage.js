import React from "react";
import ExpenseList from "./ExpenseList";
import SortingText from "./SortingText";

export default class DashboardPage extends React.Component {
  render() {
    return (
      <div>
        <p>This is a Homepage</p>
        <SortingText />
        <ExpenseList />
      </div>
    );
  }
}
