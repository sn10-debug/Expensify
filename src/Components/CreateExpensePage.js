import React from "react";
import { connect } from "react-redux";
import { addExpense } from "../actions/expenses";
import moment from "moment";
import { SingleDatePicker } from "react-dates";
import "react-dates/lib/css/_datepicker.css";

console.log(moment().format());

class CreateExpensePage extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      name: "",
      description: "",
      message: null,
      amount: "",
      date: moment(),
      calFocused: false,
    };

    this.onFormSubmit = this.onFormSubmit.bind(this);
    this.onNameChange = this.onNameChange.bind(this);
    this.onDescChange = this.onDescChange.bind(this);
    this.onAmountChange = this.onAmountChange.bind(this);
    this.onDateChange = this.onDateChange.bind(this);
    this.onFocusChange = this.onFocusChange.bind(this);
  }

  onFormSubmit(e) {
    e.preventDefault();
    let name = e.target.elements.name.value;
    let description = e.target.elements.description.value;
    let amount = e.target.elements.amount.value * 1;
    let date = new Date(this.state.date.format());

    if (!name || !description || !amount) {
      this.setState(() => ({
        message: "Make sure you fill all the details properly",
      }));
      return;
    }

    this.setState(() => {
      return { message: "Successfully Submitted" };
    });

    this.props.dispatch(addExpense({ name, description, amount, date }));
    console.log(this.props);
    this.props.history.push("/");

    // This will push to the previous page
  }

  onNameChange(e) {
    let name = e.target.value;
    this.setState(() => ({ name }));
  }

  onDescChange(e) {
    let description = e.target.value;
    this.setState(() => ({ description }));
  }

  onAmountChange(e) {
    let amount = e.target.value;
    let exp = /((^\d+(\.\d{0,2})?$)|(^[]{0}$))/;

    if (exp.test(amount)) this.setState(() => ({ amount }));
  }

  onDateChange(date) {
    if (!date) return;
    this.setState(() => ({ date }));
  }

  onFocusChange({ focused }) {
    this.setState(() => ({ calFocused: focused }));
  }

  render() {
    // numberOfMonths defines the number of months should be displayed when the user clicks on date input

    return (
      <div>
        <p>This is the page where you create Expense</p>
        <form onSubmit={this.onFormSubmit}>
          <label>Expense Name : </label>
          <input
            name="name"
            placeholder="Expense Name"
            type={"text"}
            autoFocus
            value={this.state.name}
            onChange={this.onNameChange}
          />
          <br />
          <label>Description : </label>
          <textarea
            name="description"
            placeholder="Description"
            onChange={this.onDescChange}
          ></textarea>
          <br />
          <label>Amount : </label>
          <input
            name="amount"
            type={"text"}
            placeholder="0"
            value={this.state.amount}
            onChange={this.onAmountChange}
          />
          <br />
          <SingleDatePicker
            date={this.state.date}
            onDateChange={this.onDateChange}
            focused={this.state.calFocused}
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            isOutsideRange={() => false}
          />
          <br />

          <button>Create</button>
        </form>

        {this.state.message ? <p>{this.state.message}</p> : ""}
      </div>
    );
  }
}

export default connect()(CreateExpensePage);

// The mapping function we pass into connect to map state and props , over here the first parameter is state and the second parameter is props
// and the second props consists the attributes which we pass inline while creating the component
