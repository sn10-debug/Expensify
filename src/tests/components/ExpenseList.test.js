import React from "react";
import { ExpenseList } from "../../Components/ExpenseList";
import { shallow } from "enzyme";
import { addExpense } from "../../actions/expenses";
import moment from "moment";

test("Testing the Expense List", () => {
  let wrapper = shallow(
    <ExpenseList
      expenses={[
        {
          ...addExpense({
            name: "Water Bill",
            amount: 30000,
            description: "Monthly Water bill",
          }).expense,
          id: "20",
          createDate: moment().startOf("month"),
        },
        {
          ...addExpense({
            name: "Gas Bill",
            amount: 40000,
            description: "Monthly Gas bill",
          }).expense,
          id: "20",
          createDate: moment().startOf("month"),
        },
      ]}
    />
  );
  expect(wrapper).toMatchSnapshot();
});
