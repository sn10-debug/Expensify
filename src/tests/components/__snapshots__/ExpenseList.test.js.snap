// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing the Expense List 1`] = `
<div>
  <p>
    Expense List : 
  </p>
  <React.Memo
    amount={30000}
    createDate={"2022-09-30T18:30:00.000Z"}
    description="Monthly Water bill"
    id="20"
    key="20"
    name="Water Bill"
  />
  <React.Memo
    amount={40000}
    createDate={"2022-09-30T18:30:00.000Z"}
    description="Monthly Gas bill"
    id="20"
    key="20"
    name="Gas Bill"
  />
</div>
`;
